File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/C-Постфиксная_запись/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    stack: list[int] = []
    operations: dict[str, callable[int, int]] = {
        '+': lambda x, y: x + y,
        '-': lambda x, y: x - y,
        '*': lambda x, y: x * y,
        '/': lambda x, y: x / y,
    }
    commands: list[str] = sys.stdin.readline().rstrip().split()
    for command in commands:
        if command in '+-*/':
            y, x = stack.pop(), stack.pop()
            stack.append(operations[command](x, y))
        else: stack.append(int(command))
    sys.stdout.write(f'{stack[-1]}\n')


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/C-Постфиксная_запись/solution.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    vector<int> stack;
    string simbol;
    while (cin >> simbol) {
        if (simbol == "+" || simbol == "-" || simbol == "*" || simbol == "/") {
            int x, y;
            y = stack.back();
            stack.pop_back();
            x = stack.back();
            stack.pop_back();
            stack.push_back(
                (simbol == "+"? x + y : simbol == "-"? x - y : simbol == "*"? x * y : x / y)
            );
        } else stack.push_back(stoi(simbol));
    }
    printf("%d\n", stack.back());  
}


int main() {
    solution();
}